analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs

```diff
@@ -27,20 +28,100 @@ namespace SonarAnalyzer.SymbolicExecution.Roslyn.OperationProcessors
+                        BoolConstraint.From(valueConstraint.Equals(ObjectConstraint.Null)),
+                OperationKindEx.RecursivePattern when IRecursivePatternOperationWrapper.FromOperation(pattern.WrappedOperation) is var recursivePattern =>
+                    recursivePattern switch
+                    {
+                        _ when valueConstraint.Equals(ObjectConstraint.Null) => BoolConstraint.False,
+                        {
+                            PropertySubpatterns.Length: 0,
+                            DeconstructionSubpatterns.Length: 0,
+                        } => BoolConstraint.True,
+                        {
+                            PropertySubpatterns: var propertySubPatterns,
+                            DeconstructionSubpatterns: var deconstructSubpatterns,
+                        } when propertySubPatterns // check if all sub pattern are always matching.
+                                   .Select(x => IPropertySubpatternOperationWrapper.FromOperation(x).Pattern)
+                                   .Concat(deconstructSubpatterns.Select(x => IPatternOperationWrapper.FromOperation(x)))
+                                   .All(x => x is { WrappedOperation.Kind: OperationKindEx.DiscardPattern }
+                                        || (x.WrappedOperation.Kind == OperationKindEx.DeclarationPattern
+                                            && IDeclarationPatternOperationWrapper.FromOperation(x.WrappedOperation).MatchesNull)) =>
+                                BoolConstraint.True,
+                        _ => null,
+                    },
+                OperationKindEx.DeclarationPattern when IDeclarationPatternOperationWrapper.FromOperation(pattern.WrappedOperation) is var declarationPattern =>
+                        declarationPattern switch
+                        {
+                            { MatchesNull: true } => BoolConstraint.True,
+                            { MatchesNull: false } when valueConstraint.Equals(ObjectConstraint.Null) => BoolConstraint.False,
+                            var notNull when IsTypeAssignableTo(notNull.InputType, notNull.NarrowedType) => BoolConstraint.From(valueConstraint.Equals(ObjectConstraint.NotNull)),
+                            _ => null,
+                        },
+                OperationKindEx.TypePattern when
+                    ITypePatternOperationWrapper.FromOperation(pattern.WrappedOperation) is var typePattern
+                    && IsTypeAssignableTo(typePattern.InputType, typePattern.NarrowedType) =>
+                        BoolConstraint.From(valueConstraint.Equals(ObjectConstraint.NotNull)),
+                OperationKindEx.NegatedPattern when
+                    INegatedPatternOperationWrapper.FromOperation(pattern.WrappedOperation) is var negatedPattern =>
+                        MatchValueConstraintToPattern(state, valueConstraint, negatedPattern.Pattern)?.Opposite as BoolConstraint,
```

Comment:
Unfortunately we can not use covariant return types because of the runtime. We could do this to get a bit closer, but it doesn't really bring much value:
```cs
    public abstract class SymbolicConstraint<T> where T : SymbolicConstraint<T>
    {
        public abstract SymbolicConstraint<T> Opposite { get; }
    }

    public sealed class BoolConstraint : SymbolicConstraint<BoolConstraint>
    {
        public static readonly BoolConstraint True = new();
        public static readonly BoolConstraint False = new();

        public override SymbolicConstraint<BoolConstraint> Opposite =>
            this == True ? False : True;
    }
```
I will change the signature as suggested.

Reply:
It was not about covariance. I meant to simply return `SymbolicConstraint` instead of `BoolConstraint` everywhere